docs/langref/relay_type.rst

```diff
@@ -0,0 +1,222 @@
+consists of a list of type parameters, a set of type constraints,
+a sequence of argument types, and a return type.
+
+We informally write function types as:
+`fn<type_params>(arg_types) -> ret_type where type_constraints`
+
+A type parameter in the function type may appear in the argument
+types or the return types. Additionally, each of the type constraints
+must hold at every call site of the function. The type constraints
+typically take the function's argument types and the function's return
+type as arguments, but may take a subset instead.
+
+See :py:class:`~tvm.relay.ty.FuncType` for its definition and documentation.
+
+Type Relation
+~~~~~~~~~~~~~
+
+A type relation is the most complex type system feature in Relay.
+It allows users to extend type inference with new rules.
+We use type relations to define types for operators that work with
+tensor shapes in complex ways, such as broadcasting operators or
+:code:`flatten`, allowing Relay to statically reason about the shapes
+in these cases.
+
+A type relation :code:`R` is an n-ary-input, single-output relation over
+types. Namely, :code:`R` specifies a relationship between its argument
+types and outputs `true` if the relationship holds and `false`
+if it fails to hold. Types given to a relation may be incomplete or
+include shape variables, so type inference must assign appropriate
+values to incomplete types and shape variables for necessary relations
+to hold, if such values exist.
+
+For example we can define an identity relation to be:
+
+.. code-block:: prolog
+    Identity(I, I) :- true
+
+Or we can define the relation for :code:`flatten`:
+
+.. code-block:: prolog
+    Flatten(Tensor(sh, bt), O) :-
+      O = Tensor(sh[0], prod(sh[1:]))
+
+If we have a relation like :code:`Broadcast` it becomes possible
+to type operators like :code:`add`:
+
+.. code-block:: python
+    add : fun<t1 : Type, t2 : Type, t3 : Type>(t1, t2) -> t3
+                where Broadcast
+
+The inclusion of :code:`Broadcast` above indicates that the argument
+types and the return type must be tensors where the shape of `t3` is
+the broadcast of the shapes of `t1` and `t2`. The type system will
+accept any argument types and return type so long as they fulfill
+:code:`Broadcast`. Hence, it is usually convenient to type operators
+in Relay by defining a relation specific to that operator that
+encodes all the necessary constraints on the argument types
+and the return type.
+
+Note that the above example relations are written in Prolog-like syntax,
+but currently the relations must be implemented by users in C++
+or Python. More specifically, Relay's type system uses an *ad hoc* solver
+for type relations in which type relations are actually implemented as
+C++ or Python functions that check whether the relation holds and
+imperatively update any shape variables or incomplete types. The
+functions are run as needed (if an input is updated) until one of the following holds:
```

Comment:
when is an input updated? what updates it?

Reply:
The relation's function updates it (via the type reporter, but I'm not sure if that's a detail relevant to this doc)